### ISSUE TITLE
Problem parsing assignment to parent component in a module procedure

### ISSUE THREAD

**Issue by rouson** [cmacmackin/ford#185]
The transcript below demonstrates an error message encountered when assigning to the parent component of an extended type inside a module procedure implemented in a submodule.  The code appears to be roughly the smallest example that is both standard Fortran and demonstrates the issue.  This issue appears to be related to FORD parsing the left-hand-side of the assignment to "this%parabola" and it appears to happen only when the Fortran 2008 "module procedure" statement is used in the procedure implementation to avoid repeating all of the interface information.  This was tested with FORD 5.0.4.  The code compiles with gfortran 6.2.0 or 7.0.0.

```
$ cat setter-reproducer.md 
project: constructor-issue
summary: reproducer
src_dir: .

$ cat shot.f90 
module shot_interface
  type parabola
  end type
  type, extends(parabola) :: shot
  end type
  interface
    module subroutine setter(this)
      type(shot) :: this
    end subroutine
  end interface
end module

submodule(shot_interface) shot_implementation
contains
  module procedure setter
    this%parabola = parabola()  ! assign the intrinsic structure constructor result to the parent component
  end procedure
end submodule

program main
end program

$ ford setter-reproducer.md 
Reading file shot.f90

Processing documentation comments...
Correlating information from different parts of your project...
Traceback (most recent call last):
  File "/usr/local/bin/ford", line 11, in <module>
    load_entry_point('FORD==5.0.4', 'console_scripts', 'ford')()
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/__init__.py", line 356, in run
    main(proj_data,proj_docs,md)
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/__init__.py", line 319, in main
    project.correlate()
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/fortran_project.py", line 205, in correlate
    if type(container) != str: container.correlate(self)
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/sourceform.py", line 903, in correlate
    mp.correlate(project)
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/sourceform.py", line 860, in correlate
    argname = argname or call.lower() == self.retvar.name.lower()
AttributeError: 'NoneType' object has no attribute 'name'

$ ford --version
FORD, version 5.0.4
```

**Comment by cmacmackin** [comment]
Hmm... This is a bizarrely specific bug. I'll try to fix it over the next few weeks.

**Comment by rouson** [comment]
In the process of trying to help a student who encountered this issue in his final project for the class I'm teaching, I isolated the problem a bit more.  I now think it has to do with the new use of "module procedure" that Fortran 2008 added to the language.  The previously reported error results even when with no type extension and therefore no parent component: 
```
$ cat setter-reproducer.md 
project: constructor-issue
summary: reproducer
src_dir: .

$ cat shot.f90 
module shot_interface
  type parabola
  end type
  interface
    module subroutine setter()
    end subroutine
  end interface
end module

submodule(shot_interface) shot_implementation
contains
  module procedure setter
    print *,parabola()
  end procedure
end submodule

program main
end program

$ ford setter-reproducer.md 
Reading file shot.f90

Processing documentation comments...
Correlating information from different parts of your project...
Traceback (most recent call last):
  File "/usr/local/bin/ford", line 11, in <module>
    load_entry_point('FORD==5.0.4', 'console_scripts', 'ford')()
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/__init__.py", line 356, in run 
    main(proj_data,proj_docs,md)
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/__init__.py", line 319, in main
    project.correlate()
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/fortran_project.py", line 205, in correlate
    if type(container) != str: container.correlate(self)
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/sourceform.py", line 903, in correlate
    mp.correlate(project)
  File "/usr/local/lib/python2.7/dist-packages/FORD-5.0.4-py2.7.egg/ford/sourceform.py", line 860, in correlate
    argname = argname or call.lower() == self.retvar.name.lower()
AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by szaghi** [comment]
@rouson 

O.T. Dear Damian,  are your teachings *material* somehow *accessible* (papers, examples, notes, slides...)? I am too old to be one of your students, but I feel me much like one of them...

**Comment by rouson** [comment]
@szhagi, sorry I didn't reply. Please send me an email or book a time to talk at http://rouson.youcanbook.me.

**Comment by rouson** [comment]
@cmacmackin, any updates on fixing this bug?  I just introduced Cranfield University Lecturer Salvatore Filippone (@sfilippone) to FORD and he very much liked it so we're planning to use it in a [modern Fortran course] we're co-teaching next week.   As a test case, we ran FORD on the Parallel Sparse Basic Linear Algebra Subroutines ([PSBLAS]) library that Salvatore leads.  Based on that initial test, Salvatore submitted issue #202, which is a higher priority than this issue I submitted because we'll use the PSBLAS example next week if possible, but I thought I'd check for updates on this one too.

[modern Fortran course]: https://www.cranfield.ac.uk/courses/short/aerospace/software-engineering-and-parellel-programming-in-modern-fortan
[PSBLAS]: http://people.uniroma2.it/salvatore.filippone/psblas/

**Comment by cmacmackin** [comment]
Sorry for the delay. I've been doing a lot of programming for my PhD lately, so I haven't had the mental energy to work on FORD in my own time. I was planning to go through a bunch of bug reports around now, but then my laptop broke, so I won't be able to until it's fixed. Hopefully I'll be able to start next weekend.

**Comment by cmacmackin** [comment]
Thanks for the bug report and I'm sorry that it's taken more than a year to address it. I've identified the issue. It turns out I was giving module subroutines such as this return variables, which obviously they shouldn't have. They were set to be None (a Python object with no attributes) and this was causing the error. By ensuring module subroutines are not given return variables I was able to fix the bug. This will be in the next release, likely early in the new year.
